add isochrone snapped center coordinates to response

nilsnolde opened the pull request:
#1961 

Adds the snapped center point of an isochrone to the response.

This is my first ever C++ code :champagne: Wasn't the most difficult task though. Still not sure if this is properly done:

- I'm not sure what the `valhalla::Location::path_edges` represent before the [start of the isochrone computation](https://github.com/gis-ops/valhalla/blob/7657882e776517a386a4ffa55b5a807d01dbf809/src/thor/isochrone.cc#L1051). I'd guess all edges found near the input location?! Does it then expand the search from all these edges it found nearby the input location? In which case there actually isn't smth like a single snapped point?!

- Which property of the `valhalla::Location_PathEdge::Location_PathEdge` is giving me the actual snapped point for a given edge in `valhalla::Location::path_edges`? I can see `percent_along()` as a function, suggesting it does know the snapped origin. The only attribute I found is `ll_`, but whether that's the edge geometric center, one of it's start/end nodes or the actual snapped origin, I'm not sure..

I'll amend if you'd be so nice to quickly point me in the right direction. And also apply this to the other `Compute` functions for
- [x] multi-modal.
- [x] reverse

## Tasklist

 - [x] Add tests
 - [ ] Review - you must request approval to merge any PR to master
 - [ ] Add #fixes with the issue number that this PR addresses
 - [ ] Generally use squash merge to rebase and clean comments before merging
 - [ ] Update the [changelog](CHANGELOG.md)
 - [ ] Update relevant [documentation](docs/README.md)

nilsnolde commented:
Ok, I think I got it now.. I misunderstood some of the code in the json serializer and I actually found a bug (#2115) concerning multiple locations for isochrones. Also, I added the same logic to reverse search and multimodal.

Should we address #2115 before finishing off this PR?

And my first question is answered I think: the `path_edges` for each origin are (I think) all edges connected to the node it snapped to and the `ll()` function for each edge gives the same coordinate, so I assume that's the starting node for isochrone expansion. That'd be nice and would also answer my second question:)

The build for OSX is failing with 

```
In file included from /Users/distiller/project/src/midgard/gridded_data.cc:1:
/Users/distiller/project/valhalla/midgard/gridded_data.h:4:10: fatal error: 'valhalla/proto/tripcommon.pb.h' file not found
#include "valhalla/proto/tripcommon.pb.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Not sure, what's wrong, but looks like a rookie mistake. The Azure Win build failures look like the same issue:

```
D:\a\1\s\valhalla\midgard/gridded_data.h(4): fatal error C1083: Cannot open include file: 'valhalla/proto/tripcommon.pb.h': No such file or directory [D:\a\build\src\midgard\valhalla-midgard.vcxproj]
  polyline2.cc
```

Any advice on those?

kevinkreiser commented on src/tyr/isochrone_serializer.cc:
```diff
@@ -22,12 +22,15 @@ template <class coord_t>
   int i = 0;
   auto features = array({});
+  // no support for multiple locations per request currently
+  valhalla::LatLng snapped_center = snapped_centers.at(0);
```
instead of this you can do

`request.options().locations(0).path_edges(0).ll()`

which will allow you to remove pretty much all the other code in this pr expect the serialization below.

we do have another problem though. the api supports multiple locations and we dont know to which contour a center belongs. whats worse is that, the way that we build the contours we loose all ability to do that anyway...

in my humble opinion i think we should add a new parameter to the api, that also serializes another geojson feature with geometry type MultiPoint and from there we can add all the candidate points we used to make the isochrones. @nilsnolde what would you think of that?

nilsnolde replied:
> request.options().locations(0).path_edges(0).ll()

> which will allow you to remove pretty much all the other code in this pr except the serialization below.

Oh jeez, really :sweat_smile: Thanks for the pointer, so easy..

> the api supports multiple locations

Yeah, David told me how that's supposed to work. However, that seems to be a slightly broken feature atm (IMHO), see #2115 . You think that needs to be addressed before or in general @kevinkreiser ?

> we should add a new parameter to the api, that also serializes another geojson feature with geometry type MultiPoint 

I like that better as well, then those points can be rendered directly from the FeatureCollection! I guess you don't want to hijack `show_locations` for this purpose? Rather add a new parameter, like `snapped_locations`?

kevinkreiser replied:
I think hijiacking show_locations is fine. we just need a property that distinguishes input point from snapped point and i think we would be good to go!

nilsnolde commented on src/tyr/isochrone_serializer.cc:
```diff
@@ -85,17 +85,29 @@ serializeIsochrones(const Api& request,
+        {{"type", std::string("Feature")},
+         {"properties", map({{"type", std::string("snapped")}})},
+         {"geometry", map({{"type", std::string("MultiPoint")}, {"coordinates", snapped_coords}})}}));
   }
```
I wasn't sure if should make it DRYer.. But in C++ that seems to add more complexity than it's worth it.

kevinkreiser replied:
im not sure what you mean by DYRer but what you have here looks good to me

nilsnolde replied:
Cool, thanks. I meant, it's repetitive and in most other languages I'd try to make it less repetitive;)

nilsnolde commented:
I reverted back to master and did what you suggested @kevinkreiser. I just changed the whole location response to MultiPoint and each, input & snapped, are one feature with a property indicating which is which.

![bla](https://user-images.githubusercontent.com/25637358/70815427-820c2280-1dde-11ea-8f1b-4e96428d4e43.png)

nilsnolde commented:
Small bump :innocent:

kevinkreiser commented on src/tyr/isochrone_serializer.cc:
```diff
@@ -85,17 +85,29 @@ serializeIsochrones(const Api& request,
-               {"geometry", map({{"type", std::string("Point")},
-                                 {"coordinates", array({fp_t{location.ll().lng(), 6},
-                                                        fp_t{location.ll().lat(), 6}})}})}}));
+      const valhalla::LatLng& latLng_snapped = location.path_edges(0).ll();
```
here you actually want another loop over all the path edges. there can be multiple edge candidates per input location depending on the radius you provide. so yeah just add another for loop here over `location.path_edges()`

nilsnolde replied:
But we don't want to return the `ll()` points for all candidate edges for every origin, right? I wasn't sure what it exactly means if there's multiple edges for an origin, but I figured it's when the starting node is connected to more than one edge. In the few debug tests I did I only ever saw the same `ll()` point per origin for all its candidate edges. But that's not correct then, I guess?!

Also, what exact point on an edge does `ll()` reference? If I remember right, I think I couldn't quite figure that out.

kevinkreiser replied:
i dont see why we wouldnt want to show all of the candidates that were used to make the isochrone. so when you give a lat lon to valhalla, what happens is we "intersect" that with the graphs *edges*. if you specified a radius of 100m, you'll get a bunch of snapped locations to a bunch of edges for a given location. we use *all* of these edges snapped locations to seed the expansion list that computes the isochrones. in other words, its as if you started at all those locations. if you didnt specify a radius you wont get multiple snapped locations except in the case that the closest thing to your input location was a *node* in the graph, in which case you will get a snapped location at that exact node for every edge that leaves and enters it (that is accessable to your mode of travel). i hope that explains it!

nilsnolde replied:
Thanks for the explanation @kevinkreiser . That explains why I only observed the same snapped locations for multiple candidate edges, I never specified a radius.

From a user perspective, I'd expect a single point per origin point in the case where I can't correlate the center points to origin points, as is the case currently for Valhalla as you pointed out earlier. I just Imagine origins being close together would otherwise potentially have overlapping clusters of center points. Does that make sense?

nilsnolde replied:
Maybe return the gravity point of center point clusters?

kevinkreiser commented on src/tyr/isochrone_serializer.cc:
```diff
@@ -85,17 +85,29 @@ serializeIsochrones(const Api& request,
+
+    features->emplace_back(map(
+        {{"type", std::string("Feature")},
+         {"properties", map({{"type", std::string("snapped")}})},
```
if each point was an individual feature we could also specify the index of the location the input and snap points belonged to. it would make the response more verbose but, ultimately more useful.

nilsnolde replied:
Oops, just reading this now. Actually that's true, that way we could correlate snapped to input points! Good stuff.

I'll do this as soon as I got Valhalla set up again. Had to wipe my Ubuntu a few days ago and went with Arch now, will have to dive more into prime-server deps, didn't work on first try. Should be done after the weekend.

nilsnolde commented on src/tyr/isochrone_serializer.cc:
```diff
@@ -85,17 +85,43 @@ serializeIsochrones(const Api& request,
+          snapped_points_array->push_back(
+              array({fp_t{current_coords.at(0), 6}, fp_t{current_coords.at(1), 6}}));
+        }
+      };
```
The `current_coords` vector is just so I can remove duplicates for the case when the snapped object is a node. Don't really like that, so any improvement suggestion is welcome.

kevinkreiser replied:
yep there is a common c++ pattern you can use here where you insert it into a "hash set" and when you do the insertion you check if it actually got inserted (meaning the set hadn't see this unique point before). here is some pseudo code:

```c++
std::unordered_set<midgard::PointLL> snaps;
for(auto loc : request.locs()){
  for(auto pe : loc.path_edges()) {
    if(snaps.insert(midgard::PointLL(pe.lng(), pe.lat())).second) {
      //add feature as we havent seen this snap point before
    }
  }
}
```

nilsnolde replied:
Ok, thx, I changed accordingly.

nilsnolde commented:
ok, I think now it *could* be good to go @kevinkreiser :) Each origin gets one input `Point` feature and multiple aggregated snapped points as `MultiPoint` feature and both have `location_index` as additional property.

nilsnolde commented:
Damn, smth went wrong with merging master.. Might be a good idea to squash the commits, should be ok other than that. Next time will be smoother!

nilsnolde commented:
What do you think @kevinkreiser ? Can we merge this?

kevinkreiser commented:
@nilsnolde thanks for updating the PR. when the builds pass we'll merge it
